src/viam/logging.py

```diff
@@ -1,11 +1,43 @@
+import asyncio
 import logging
 import sys
 from copy import copy
-from logging import DEBUG, ERROR, FATAL, INFO, WARN, WARNING  # noqa: F401
+from datetime import datetime
+from logging import DEBUG, ERROR, FATAL, INFO, WARN, WARNING, LogRecord  # noqa: F401
 from typing import Dict
 
+import viam
+
 LOG_LEVEL = INFO
 LOGGERS: Dict[str, logging.Logger] = {}
+_MODULE_PARENT = None
+
+
+class ModuleHandler(logging.Handler):
+    def __init__(self, logger: logging.Logger):
+        self.logger = logger
+        self.parent = _MODULE_PARENT
+
+        logger.handlers.clear()
+        super().__init__()
+
+    def emit(self, record: LogRecord):
+        assert isinstance(record, logging.LogRecord)
+        stack = f"exc_info: {record.exc_info}, exc_text: {record.exc_text}, stack_info: {record.stack_info}"
+        message = f"{record.filename}:{record.lineno}\t{record.msg}"
+        time = datetime.fromtimestamp(record.created)
+
+        try:
+            assert self.parent is not None
+            asyncio.create_task(
+                self.parent.log(record.name, record.levelname, time, message, stack), name=f"{viam._TASK_PREFIX}-LOG-{time}"
+            )
+        except Exception as err:
+            self.logger.handlers.clear()
+            addHandlers(self.logger, True)
+            self.logger.error(f"Module handler failed for {record.name} - {err}")
+            self.logger.log(record.levelno, message)
```

Comment:
I'm not sure how to make the logs appear like they're still from the original source. For example:
```
\_ 2024-03-08 17:16:11,222		INFO	src.gizmo.my_gizmo (logging.py:41)	my_gizmo.py:79	gizmo1 is closed.
```
Here, it says that the log is from `logging.py:41`. For now, I made it so that the message includes the filename and number it's from

Reply:
This log already looks solid when you test it with the RDK! I just ran the RDK with the complex module example code and a `custom_run.sh` as we discussed offline; follow up with me if you want to test it the same way I did.

```
2024-03-10T22:26:43.009-0400	INFO	robot_server.src.gizmo.my_gizmo	my_gizmo.py:83	gizmo1 is closed.	{"log_ts": "2024-03-10T22:26:43.007Z"}
```

Reply:
The log you see is if the log succeeded. But when it fails, we still need to log, which is what this `except` is doing. The log line I posted above was while testing that failed logs still show up using `stdout` and `stderr`. But because it's calling the log from `logging.py`, it comes up as `logging.py:41` as the filename and line number

Reply:
Gotcha... Let me think about that one. I think it's definitely an edge case that we fail to log over gRPC once gRPC logging has started, so unclear how "bad" it is that we report two different line numbers.